File: bindings/fsharp/src/STRling/Emitters/Pcre2.fs
namespace STRling.Emitters

open System
open System.Text
open STRling
open STRling.Core

/// PCRE2 Emitter - IR to PCRE2 Pattern String
/// This module implements the emitter that transforms STRling's Intermediate
/// Representation (IR) into PCRE2-compatible regex pattern strings.
module Pcre2 =
    
    /// Characters that need escaping in PCRE2 patterns.
    let private metaChars = set ['.'; '^'; '$'; '|'; '('; ')'; '?'; '*'; '+'; '{'; '}'; '['; ']'; '\\']
    
    /// Escape PCRE2 metacharacters in literal strings.
    let escapeLiteral (s: string) : string =
        let sb = StringBuilder()
        for i = 0 to s.Length - 1 do
            let ch = s.[i]
            if metaChars.Contains(ch) then
                sb.Append('\\').Append(ch) |> ignore
            else
                sb.Append(ch) |> ignore
        sb.ToString()
    
    /// Escape a character for use inside [...] per PCRE2 rules.
    let escapeClassChar (ch: char) : string =
        match ch with
        | '\\' -> "\\\\"
        | ']' -> "\\]"
        | '[' -> "\\["
        | '-' -> "\\-"
        | '^' -> "\\^"
        | '\n' -> "\\n"
        | '\r' -> "\\r"
        | '\t' -> "\\t"
        | '\f' -> "\\f"
        | '\u000B' -> "\\v"
        | c when int c < 32 || (int c >= 127 && int c <= 159) ->
            sprintf "\\x%02x" (int c)
        | c -> string c
    
    /// Emit a PCRE2 character class.
    let private emitClass (negated: bool) (items: IRClassItem list) : string =
        match items with
        | [IRClassEscape (k, prop)] ->
            match k, negated, prop with
            | "d", true, _ -> "\\D"
            | "w", true, _ -> "\\W"
            | "s", true, _ -> "\\S"
            | "d", false, _ -> "\\d"
            | "w", false, _ -> "\\w"
            | "s", false, _ -> "\\s"
            | "D", true, _ -> "\\d"
            | "W", true, _ -> "\\w"
            | "S", true, _ -> "\\s"
            | "D", false, _ -> "\\D"
            | "W", false, _ -> "\\W"
            | "S", false, _ -> "\\S"
            | ("p" | "P"), _, Some propVal ->
                let isUpperP = (k = "P")
                let useUpperP = negated <> isUpperP
                let useK = if useUpperP then "P" else "p"
                sprintf "\\%s{%s}" useK propVal
            | _ ->
                "[" + (if negated then "^" else "") + "\\" + k + "]"
        | _ ->
            let parts = StringBuilder()
            for item in items do
                match item with
                | IRClassLiteral c ->
                    parts.Append(escapeClassChar (c.[0])) |> ignore
                | IRClassRange (f, t) ->
                    parts.Append(escapeClassChar (f.[0])).Append('-').Append(escapeClassChar (t.[0])) |> ignore
                | IRClassEscape (typ, propOpt) ->
                    match typ, propOpt with
                    | "d", _ -> parts.Append("\\d") |> ignore
                    | "D", _ -> parts.Append("\\D") |> ignore
                    | "w", _ -> parts.Append("\\w") |> ignore
                    | "W", _ -> parts.Append("\\W") |> ignore
                    | "s", _ -> parts.Append("\\s") |> ignore
                    | "S", _ -> parts.Append("\\S") |> ignore
                    | ("p" | "P"), Some prop ->
                        parts.Append('\\').Append(typ).Append('{').Append(prop).Append('}') |> ignore
                    | _, _ ->
                        parts.Append('\\').Append(typ) |> ignore
            "[" + (if negated then "^" else "") + parts.ToString() + "]"
    
    /// Emit *, +, ?, {m}, {m,}, {m,n} plus optional lazy/possessive suffix.
    let private emitQuantSuffix (minv: int) (maxv: string) (mode: string) : string =
        let q =
            match minv, maxv with
            | 0, "Inf" -> "*"
            | 1, "Inf" -> "+"
            | 0, "1" -> "?"
            | n, m when m <> "Inf" && string n = m -> sprintf "{%d}" n
            | n, "Inf" -> sprintf "{%d,}" n
            | n, m -> sprintf "{%d,%s}" n m
        match mode with
        | "Lazy" -> q + "?"
        | "Possessive" -> q + "+"
        | _ -> q
    
    /// Return true if 'child' needs a non-capturing group when quantifying.
    let private needsGroupForQuant (child: IROp) : bool =
        match child with
        | IRCharClass _ -> false
        | IRDot -> false
        | IRGroup _ -> false
        | IRBackref _ -> false
        | IRAnchor _ -> false
        | IRLit v ->
            if v.Length = 2 && v.[0] = '\\' then false
            else v.Length > 1
        | IRAlt _ -> true
        | IRLook _ -> true
        | IRSeq parts -> parts.Length > 1
        | IRQuant _ -> true
    
    /// Generate opening for group based on type.
    let private emitGroupOpen (capturing: bool) (name: string option) (atomic: bool) : string =
        if atomic then "(?>"
        elif capturing then
            match name with
            | Some n -> sprintf "(?<%s>" n
            | None -> "("
        else "(?:"
    
    /// Emit a single IR node to PCRE2 syntax.
    let rec private emitNode (parentKind: string) (node: IROp) : string =
        match node with
        | IRLit v -> escapeLiteral v
        | IRDot -> "."
        | IRAnchor at ->
            match at with
            | "Start" -> "^"
            | "End" -> "$"
            | "WordBoundary" -> "\\b"
            | "NotWordBoundary" -> "\\B"
            | "NonWordBoundary" -> "\\B"
            | "AbsoluteStart" -> "\\A"
            | "EndBeforeFinalNewline" -> "\\Z"
            | "AbsoluteEnd" -> "\\z"
            | _ -> ""
        | IRBackref (byIndex, byName) ->
            match byName with
            | Some n -> sprintf "\\k<%s>" n
            | None ->
                match byIndex with
                | Some i -> "\\" + string i
                | None -> ""
        | IRCharClass (negated, items) -> emitClass negated items
        | IRSeq parts -> parts |> List.map (emitNode "Seq") |> String.concat ""
        | IRAlt branches ->
            let body = branches |> List.map (emitNode "Alt") |> String.concat "|"
            if parentKind = "Seq" || parentKind = "Quant" then "(?:" + body + ")"
            else body
        | IRQuant (child, min, max, mode) ->
            let childStr = emitNode "Quant" child
            let grouped = 
                match child with
                | IRGroup _ -> childStr
                | _ when needsGroupForQuant child -> "(?:" + childStr + ")"
                | _ -> childStr
            grouped + emitQuantSuffix min max mode
        | IRGroup (capturing, body, name, atomic) ->
            emitGroupOpen capturing name atomic + emitNode "Group" body + ")"
        | IRLook (dir, neg, body) ->
            let op =
                match dir, neg with
                | "Ahead", false -> "?="
                | "Ahead", true -> "?!"
                | "Behind", false -> "?<="
                | "Behind", true -> "?<!"
                | _, _ -> "?="
            "(" + op + emitNode "Look" body + ")"
    
    /// Build the inline prefix form expected by tests, e.g. "(?imx)".
    let private emitPrefixFromFlags (flags: Flags) : string =
        let sb = StringBuilder()
        if flags.IgnoreCase then sb.Append('i') |> ignore
        if flags.Multiline then sb.Append('m') |> ignore
        if flags.DotAll then sb.Append('s') |> ignore
        if flags.Unicode then sb.Append('u') |> ignore
        if flags.Extended then sb.Append('x') |> ignore
        if sb.Length > 0 then sprintf "(?%s)" (sb.ToString())
        else ""
    
    /// Emit a PCRE2 pattern string from IR.
    let emit (irRoot: IROp) (flags: Flags option) : string =
        let prefix = 
            match flags with
            | Some f -> emitPrefixFromFlags f
            | None -> ""
        let body = emitNode "" irRoot
        prefix + body
    
    /// Emit a PCRE2 pattern string from IR without flags.
    let emitNoFlags (irRoot: IROp) : string =
        emit irRoot None

File: bindings/dart/lib/src/emitters/pcre2.dart
/// STRling PCRE2 Emitter
///
/// Transforms STRling IR into PCRE2-compatible regex strings.
/// Iron Law: Emitters are pure functions with signature emit(ir, flags) → string.

import '../core/parser.dart' show Flags;

/// Special characters that need escaping in PCRE2
const _literalSpecial = r'[\]^$.|?*+(){}';

/// Special characters inside character class
const _classSpecial = r'[\]^-';

/// Escape a literal string for use outside character classes
String _escapeLiteral(String s) {
  final buf = StringBuffer();
  for (final ch in s.split('')) {
    if (_literalSpecial.contains(ch)) {
      buf.write('\\');
    }
    buf.write(ch);
  }
  return buf.toString();
}

/// Escape a character for use inside character classes
String _escapeClassChar(String ch) {
  if (_classSpecial.contains(ch)) {
    return '\\$ch';
  }
  return ch;
}

/// PCRE2 Emitter class
class Pcre2Emitter {
  /// Emit PCRE2 pattern from IR
  ///
  /// [ir] is a Map representation of the STRling IR
  /// [flags] optional compilation flags
  /// Returns the compiled PCRE2 regex string
  String emit(Map<String, dynamic> ir, [Flags? flags]) {
    final irType = ir['ir'] as String;

    switch (irType) {
      case 'Lit':
        return _emitLit(ir);
      case 'Seq':
        return _emitSeq(ir);
      case 'Alt':
        return _emitAlt(ir);
      case 'Group':
        return _emitGroup(ir);
      case 'Quant':
        return _emitQuant(ir);
      case 'CharClass':
        return _emitCharClass(ir);
      case 'Anchor':
        return _emitAnchor(ir);
      case 'Dot':
        return '.';
      case 'Backref':
        return _emitBackref(ir);
      case 'Look':
        return _emitLook(ir);
      case 'Esc':
        return _emitEsc(ir);
      default:
        throw FormatException('Unknown IR type: $irType');
    }
  }

  String _emitLit(Map<String, dynamic> ir) {
    final value = ir['value'] as String;
    return _escapeLiteral(value);
  }

  String _emitSeq(Map<String, dynamic> ir) {
    final parts = ir['parts'] as List;
    return parts.map((p) => emit(p as Map<String, dynamic>)).join();
  }

  String _emitAlt(Map<String, dynamic> ir) {
    final branches = ir['branches'] as List;
    return branches.map((b) => emit(b as Map<String, dynamic>)).join('|');
  }

  String _emitGroup(Map<String, dynamic> ir) {
    final body = emit(ir['body'] as Map<String, dynamic>);
    final capturing = ir['capturing'] as bool? ?? false;
    final name = ir['name'] as String?;
    final atomic = ir['atomic'] as bool? ?? false;

    if (atomic) {
      return '(?>$body)';
    }
    if (name != null) {
      return '(?<$name>$body)';
    }
    if (capturing) {
      return '($body)';
    }
    return '(?:$body)';
  }

  String _emitQuant(Map<String, dynamic> ir) {
    final child = ir['child'] as Map<String, dynamic>;
    final min = ir['min'] as int;
    final max = ir['max'];
    final mode = ir['mode'] as String? ?? 'Greedy';

    var childStr = emit(child);

    // Wrap if needed (sequences, alternations, multi-char literals)
    final needsParens = _needsQuantifierParens(child, childStr);
    if (needsParens) {
      childStr = '(?:$childStr)';
    }

    // Build quantifier suffix
    String quantStr;
    if (max == 'Inf' || max == null) {
      if (min == 0) {
        quantStr = '*';
      } else if (min == 1) {
        quantStr = '+';
      } else {
        quantStr = '{$min,}';
      }
    } else if (min == max) {
      if (min == 0) {
        return ''; // Matches nothing, effectively empty
      } else if (min == 1) {
        quantStr = '';
      } else {
        quantStr = '{$min}';
      }
    } else if (min == 0 && max == 1) {
      quantStr = '?';
    } else {
      quantStr = '{$min,$max}';
    }

    // Add mode suffix
    if (mode == 'Lazy') {
      quantStr += '?';
    } else if (mode == 'Possessive') {
      quantStr += '+';
    }

    return '$childStr$quantStr';
  }

  bool _needsQuantifierParens(Map<String, dynamic> child, String childStr) {
    final irType = child['ir'] as String;
    switch (irType) {
      case 'Seq':
        return true;
      case 'Alt':
        return true;
      case 'Lit':
        return childStr.length > 1 && !childStr.startsWith('\\');
      case 'Quant':
        return true;
      default:
        return false;
    }
  }

  String _emitCharClass(Map<String, dynamic> ir) {
    final negated = ir['negated'] as bool? ?? false;
    final items = ir['items'] as List;

    // Single-item shorthand optimization
    if (items.length == 1) {
      final item = items[0] as Map<String, dynamic>;
      final itemIr = item['ir'] as String;
      
      if (itemIr == 'Esc') {
        final type = item['type'] as String;
        
        // Handle d, w, s with negation flipping
        if ('dws'.contains(type)) {
          if (negated) {
            return '\\${type.toUpperCase()}';
          }
          return '\\$type';
        }
        
        // Handle D, W, S
        if ('DWS'.contains(type)) {
          if (negated) {
            return '\\${type.toLowerCase()}';
          }
          return '\\$type';
        }
        
        // Handle \p{...} and \P{...}
        if (type == 'p' || type == 'P') {
          final prop = item['property'] as String?;
          if (prop != null) {
            final shouldNegate = negated != (type == 'P');
            final use = shouldNegate ? 'P' : 'p';
            return '\\$use{$prop}';
          }
        }
      }
    }

    // Build bracket class
    final parts = <String>[];
    var hasHyphen = false;

    for (final item in items) {
      final itemMap = item as Map<String, dynamic>;
      final itemIr = itemMap['ir'] as String;

      switch (itemIr) {
        case 'Char':
          final ch = itemMap['char'] as String;
          if (ch == '-') {
            hasHyphen = true;
          } else {
            parts.add(_escapeClassChar(ch));
          }
          break;
        case 'Range':
          final from = itemMap['from'] as String;
          final to = itemMap['to'] as String;
          parts.add('${_escapeClassChar(from)}-${_escapeClassChar(to)}');
          break;
        case 'Esc':
          final type = itemMap['type'] as String;
          final prop = itemMap['property'] as String?;
          if (prop != null) {
            parts.add('\\$type{$prop}');
          } else {
            parts.add('\\$type');
          }
          break;
      }
    }

    // Hyphen at start to avoid ambiguity
    final inner = hasHyphen ? '-${parts.join()}' : parts.join();
    return '[${negated ? '^' : ''}$inner]';
  }

  String _emitAnchor(Map<String, dynamic> ir) {
    final at = ir['at'] as String;
    switch (at) {
      case 'Start':
        return '^';
      case 'End':
        return r'$';
      case 'WordBoundary':
        return r'\b';
      case 'NotWordBoundary':
        return r'\B';
      case 'AbsoluteStart':
        return r'\A';
      case 'AbsoluteEnd':
        return r'\z';
      case 'EndBeforeFinalNewline':
        return r'\Z';
      default:
        throw FormatException('Unknown anchor type: $at');
    }
  }

  String _emitBackref(Map<String, dynamic> ir) {
    final byIndex = ir['byIndex'] as int?;
    final byName = ir['byName'] as String?;

    if (byName != null) {
      return '\\k<$byName>';
    }
    if (byIndex != null) {
      return '\\$byIndex';
    }
    throw FormatException('Backref must have byIndex or byName');
  }

  String _emitLook(Map<String, dynamic> ir) {
    final dir = ir['dir'] as String;
    final neg = ir['neg'] as bool? ?? false;
    final body = emit(ir['body'] as Map<String, dynamic>);

    if (dir == 'Ahead') {
      return neg ? '(?!$body)' : '(?=$body)';
    } else {
      return neg ? '(?<!$body)' : '(?<=$body)';
    }
  }

  String _emitEsc(Map<String, dynamic> ir) {
    final type = ir['type'] as String;
    final prop = ir['property'] as String?;

    if (prop != null) {
      return '\\$type{$prop}';
    }
    return '\\$type';
  }
}

/// Convenience function to emit PCRE2 from IR
String emitPcre2(Map<String, dynamic> ir, [Flags? flags]) {
  return Pcre2Emitter().emit(ir, flags);
}
